Project page

KristinaGru opened the pull request:
Add Project Task List View, with possibility to add new task with a title, change tasks completed status with a checkbox and drag and drop tasks to a different list, or to reorder in the same list. Using mock data.

KristinaGru committed:
add project overview element

KristinaGru committed:
add ProjectListView and ListTable components

KristinaGru committed:
add add new task functionality

KristinaGru committed:
fix styles

KristinaGru committed:
Merge branch 'dev' of github.com:HackYourFuture-CPH/class-25-lothbrok into project-page

KristinaGru committed:
add drag and drop between lists

KristinaGru committed:
fix strict typescript and dnd library compatibility issue

KristinaGru committed:
add list reorder functionality

kralle333 commented:
Hi I see you have a conflicting file with dev. Can you merge dev into your branch and resolve the conflict?

KristinaGru committed:
Merge branch 'dev' of github.com:HackYourFuture-CPH/class-25-lothbrok into project-page

KristinaGru committed:
add /project/:id path

kralle333 commented on frontend/src/components/projectListView/ProjectListView.tsx:
```diff
@@ -0,0 +1,252 @@
+          return tasks;
+        });
+      };
+      if (destination.droppableId === 'documentation') {
+        changeOrder(setDocumentation, documentation);
+      } else if (destination.droppableId === 'ongoing') {
+        changeOrder(setOngoing, ongoing);
+      } else if (destination.droppableId === 'to_do') {
+        changeOrder(setTodo, todo);
+      } else if (destination.droppableId === 'done') {
+        changeOrder(setDone, done);
+      }
```
Use a switch case here instead. Also create constants for these different `droppableId`s
Something like:
`
const droppableIdDocumentation = 'documentation'
const droppableIdOngoing = 'ongoing'
so on...
`

kralle333 commented on frontend/src/components/projectListView/ProjectListView.tsx:
```diff
@@ -0,0 +1,252 @@
+const ProjectListView = ({ tasks }: { tasks: Task[] }) => {
+  const [allTasks, setAllTasks] = useState<Task[]>(tasks);
+
+  const [documentation, setDocumentation] = useState<Task[]>([]);
+  const [ongoing, setOngoing] = useState<Task[]>([]);
+  const [todo, setTodo] = useState<Task[]>([]);
+  const [done, setDone] = useState<Task[]>([]);
+
+  const [documentationDescription, setDocumentationDescription] =
+    useState<string>('');
+  const [ongoingDescription, setOngoingDescription] = useState<string>('');
+  const [todoDescription, setTodoDescription] = useState<string>('');
+  const [doneDescription, setDoneDescription] = useState<string>('');
+
+  const [documentationEditing, setDocumentationEditing] =
+    useState<boolean>(false);
+  const [ongoingEditing, setOngoingEditing] = useState<boolean>(false);
+  const [todoEditing, setTodoEditing] = useState<boolean>(false);
+  const [doneEditing, setDoneEditing] = useState<boolean>(false);
```
Here is a lot of repetition, you can probably abstract this away in an object. Let's say you have an object representing a Category, that can have:
- An array of tasks
- A description
- Editing status

Abstracting it in this way will then also make it possible to remove/add categories as well open up for the ability for the user to create their own categories.

kralle333 commented on frontend/src/components/projectListView/ProjectListView.tsx:
```diff
@@ -0,0 +1,252 @@
+        }
+      });
+    };
+    setDocumentation(filterTaskByStatus('documentation'));
+    setOngoing(filterTaskByStatus('ongoing'));
+    setTodo(filterTaskByStatus('to_do'));
+    setDone(filterTaskByStatus('done'));
```
Same as the comment I made elsewhere, make these values into constant and use them instead of raw string values.

kralle333 commented on frontend/src/components/projectListView/ProjectListView.tsx:
```diff
@@ -0,0 +1,252 @@
+  return (
+    <DragDropContext onDragEnd={onDragEnd}>
+      <div className="section-title">
+        <h4>Documentation</h4>
+        {documentationEditing ? (
+          <form
+            onSubmit={(e) => {
+              e.preventDefault();
+              addNewTask(
+                setDocumentation,
+                documentationDescription,
+                documentation,
+                'documentation'
+              );
+              setDocumentationEditing(false);
+              setDocumentationDescription('');
+            }}>
+            <TextField
+              placeholder="New Task Title"
+              variant="filled"
+              value={documentationDescription}
+              onChange={(e) => setDocumentationDescription(e.target.value)}
```
When you have created the category object you should be able to loop over your category types to reduce the repetition here.

kralle333 commented on frontend/src/firebase_config.js:
```diff
@@ -1,17 +1,15 @@
-    storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
-    messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
-    appId: process.env.REACT_APP_FIREBASE_APP_ID,
-    measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENT_ID,
+  apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
+  authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
+  projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
+  storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
+  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
+  appId: process.env.REACT_APP_FIREBASE_APP_ID,
+  measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENT_ID
 };
 
-
 const app = initializeApp(firebaseConfig);
-export const auth= getAuth(app);
\ No newline at end of file
+export const auth = getAuth(app);
```
I assume it is some formatter that are responsible for these changes?

kralle333 commented on frontend/src/pages/projectView/ProjectView.tsx:
```diff
@@ -0,0 +1,75 @@
+import './projectView.css';
+import ProjectListView from '../../components/projectListView/ProjectListView';
+
+export type Task = {
+  id: number | string;
+  description: string;
+  status: string;
+  due_date: string;
+  assignee: string;
+  completed: boolean;
+  priority: string;
+  project_id: number | undefined;
+  user_id: number | string;
+};
+
+type Project = {
+  id: number;
+  title: string;
+  description: string;
+  thumbnail_link: string;
+  date_of_creation: string;
+  amount_of_tasks: number;
+  user_id: number;
+};
```
I assume these objects are going to be used elsewhere, maybe it is better to extract them out into their own files

kralle333 commented on frontend/src/pages/projectView/ProjectView.tsx:
```diff
@@ -0,0 +1,75 @@
+  useEffect(() => {
+    if (id) {
+      setTasks(allTasks.filter((task) => task.project_id === +id));
+      setProject(projects.filter((project) => project.id === +id)[0]);
```
how should it act when there is no project?

kralle333 reviewed (changes requested):
All in all, great job!

I have some smaller comments that will help make the code easier to modify in the future and easier to read.

@YurijPaharev Wanna have a look to see if my comments are sensible or if I missed something important?

KristinaGru commented:
Thanks for the review Kristian! I'm on it 🙌

KristinaGru committed:
Merge branch 'dev' of github.com:HackYourFuture-CPH/class-25-lothbrok into project-page

YurijPaharev commented:
@kralle333 You did an amazing and a very thorough job, I wouldn't have done better myself

KristinaGru committed:
change from if to switch case in changeOrder func

KristinaGru committed:
refactor ProjectListView component

KristinaGru committed:
create directory for typescript types

KristinaGru committed:
fix reordering lists

kralle333 commented on frontend/src/components/projectListView/ProjectListView.tsx:
```diff
@@ -1,41 +1,29 @@
-  const [doneEditing, setDoneEditing] = useState<boolean>(false);
+  const [description, setDescription] = useState<string>('');
+  const [editing, setEditing] = useState<string>('');
+  const categories = {
+    documentation: 'documentation',
+    ongoing: 'ongoing',
+    todo: 'to_do',
+    done: 'done',
+  };
```
I am not really sure what the reason is for this map...

kralle333 commented on frontend/src/components/projectListView/ProjectListView.tsx:
```diff
@@ -79,172 +58,79 @@ const ProjectListView = ({ tasks }: { tasks: Task[] }) => {
+        <React.Fragment key={section}>
+          <div className='section-title'>
+            <h4>
+              {section === categories.documentation
+                ? 'Documentation'
+                : section === categories.ongoing
+                ? 'Ongoing'
+                : section === categories.todo
+                ? 'Pending'
+                : section === categories.done
+                ? 'Done'
+                : section}
```
If you convert your categories map to an array and then have the values be the Title, then it would make this code much simpler

kralle333 reviewed (changes requested):
Code looks much cleaner, very nice!

Last comments is about the use of a map to store the values of the categories.

KristinaGru committed:
switch to object entries loop

KristinaGru commented:
I kept categories in an object instead of an array to have a pair of list titles I use in the UI and a string that matches the task status in the database, hope that's okay. But I switched to an Object.entries loop and removed the redundant ternary operator that way 🤞.

kralle333 reviewed (approved):
Great work :) Keep it up
